File: fabric-audit-agent/core/attribution.js
const DEFAULT_TOP_N = 3;

/**
 * Decide which user(s) are driving an item's CU consumption, from that item's
 * activity records in the window. Pure: no I/O.
 *
 * Ranking — "both if available": if ANY event carries a cost proxy (cpuMs or
 * durationMs from Log Analytics), rank users by summed cost ('cost' mode);
 * otherwise rank by operation count ('frequency' mode, e.g. Activity Events only).
 *
 * Interactive vs background: each event says whether it was interactive. If the
 * item's load is mostly background (scheduled refresh, etc.), `background` is true
 * so callers name the owner/initiator — NOT an interactive "consumer."
 *
 * @param {Array<{user?:string, at?:string, interactive?:boolean, cpuMs?:number, durationMs?:number, op?:string}>} events  events for ONE item, already windowed
 * @param {{topN?:number, owner?:string}} [opts]
 * @returns {{mode:'cost'|'frequency', userCount:number, background:boolean, owner:string|null, topUsers:Array<{user:string, ops:number, cpuMs:number, interactive:boolean, score:number}>}}
 */
export function attributeUsers(events = [], opts = {}) {
  const topN = opts.topN ?? DEFAULT_TOP_N;
  const hasCost = events.some(e => Number.isFinite(e.cpuMs) || Number.isFinite(e.durationMs));
  const mode = hasCost ? 'cost' : 'frequency';

  const by = new Map();
  let bgContribution = 0;
  let totalContribution = 0;
  for (const e of events) {
    const user = String(e.user ?? '').trim();
    if (!user) continue;
    const cost = Number.isFinite(e.cpuMs) ? e.cpuMs : (Number.isFinite(e.durationMs) ? e.durationMs : 0);
    const contribution = mode === 'cost' ? cost : 1;   // weight "background" by COST in cost mode
    totalContribution += contribution;
    if (!e.interactive) bgContribution += contribution;
    const cur = by.get(user) ?? { user, ops: 0, cpuMs: 0, interactiveOps: 0 };
    cur.ops += 1;
    cur.cpuMs += cost;
    if (e.interactive) cur.interactiveOps += 1;
    by.set(user, cur);
  }

  const users = [...by.values()].map(u => ({
    user: u.user,
    ops: u.ops,
    cpuMs: Math.round(u.cpuMs),
    interactive: u.interactiveOps > 0,
    score: mode === 'cost' ? u.cpuMs : u.ops,
  }));
  // rank by score, then ops, then name (stable + deterministic)
  users.sort((a, b) => b.score - a.score || b.ops - a.ops || a.user.localeCompare(b.user));

  const background = totalContribution > 0 && bgContribution / totalContribution >= 0.5;

  return {
    mode,
    userCount: by.size,
    background,
    owner: opts.owner ?? null,
    topUsers: users.slice(0, topN),
  };
}

/**
 * Attach attribution to each item: sets topUsers / userCount / background / owner /
 * attributionMode on items that have events. Items without events are returned
 * unchanged (the detector then says "pending correlation").
 * @param {object[]} items
 * @param {Record<string, object[]>} eventsByItem  keyed by item name (or id)
 * @param {{topN?:number, owner?:string}} [opts]
 */
export function enrichItems(items = [], eventsByItem = {}, opts = {}) {
  return items.map(it => {
    const events = eventsByItem[it.name] ?? eventsByItem[it.id] ?? [];
    if (!events.length) return it;
    const a = attributeUsers(events, { topN: opts.topN, owner: it.owner ?? opts.owner });
    return {
      ...it,
      topUsers: a.topUsers,
      userCount: a.userCount,
      background: a.background,
      owner: a.owner ?? it.owner ?? null,
      attributionMode: a.mode,
    };
  });
}

File: fabric-audit-agent-py/fabric_audit_agent/attribution.py
"""User-attribution engine. Faithful port of the Node ``core/attribution.js``.

Decides which user(s) are driving an item's CU consumption from that item's activity
records in the window. Pure: no I/O.

Ranking — "both if available": if any event carries a cost proxy (``cpuMs`` or
``durationMs`` from Log Analytics), rank users by summed cost ("cost" mode); otherwise
rank by operation count ("frequency" mode, e.g. Activity Events only).

Interactive vs background: ``background`` is computed COST-weighted in cost mode, so a
single heavy refresh (high cost, low op-count) is still flagged background -> the
owner/initiator is named instead of an interactive "consumer".
"""

import math

DEFAULT_TOP_N = 3


def _is_num(v):
    # mirrors JS Number.isFinite: rejects bool, NaN, and Infinity
    return isinstance(v, (int, float)) and not isinstance(v, bool) and math.isfinite(v)


def attribute_users(events=None, *, top_n=DEFAULT_TOP_N, owner=None):
    """Rank the users driving one item's CU from its windowed events.

    :returns: ``{mode, userCount, background, owner, topUsers:[{user, ops, cpuMs, interactive, score}]}``
    """
    events = events or []
    has_cost = any(_is_num(e.get("cpuMs")) or _is_num(e.get("durationMs")) for e in events)
    mode = "cost" if has_cost else "frequency"

    by = {}
    bg_contribution = 0.0
    total_contribution = 0.0
    for e in events:
        user = str(e.get("user") or "").strip()
        if not user:
            continue
        cpu, dur = e.get("cpuMs"), e.get("durationMs")
        cost = cpu if _is_num(cpu) else (dur if _is_num(dur) else 0)
        contribution = cost if mode == "cost" else 1
        total_contribution += contribution
        if not e.get("interactive"):
            bg_contribution += contribution
        cur = by.get(user) or {"user": user, "ops": 0, "cpuMs": 0, "interactiveOps": 0}
        cur["ops"] += 1
        cur["cpuMs"] += cost
        if e.get("interactive"):
            cur["interactiveOps"] += 1
        by[user] = cur

    users = [
        {
            "user": u["user"], "ops": u["ops"], "cpuMs": int(math.floor(u["cpuMs"] + 0.5)),  # JS Math.round (half-up); cpuMs >= 0
            "interactive": u["interactiveOps"] > 0,
            "score": u["cpuMs"] if mode == "cost" else u["ops"],
        }
        for u in by.values()
    ]
    # rank by score, then ops, then name (stable + deterministic)
    users.sort(key=lambda u: (-u["score"], -u["ops"], u["user"]))

    background = total_contribution > 0 and (bg_contribution / total_contribution) >= 0.5

    return {
        "mode": mode,
        "userCount": len(by),
        "background": background,
        "owner": owner,
        "topUsers": users[:top_n],
    }


def enrich_items(items=None, events_by_item=None, *, top_n=DEFAULT_TOP_N, owner=None):
    """Attach attribution to each item that has events; leave the rest untouched.

    Sets ``topUsers / userCount / background / owner / attributionMode`` on matched items.
    ``events_by_item`` is keyed by item name (or id).
    """
    items = items or []
    events_by_item = events_by_item or {}
    out = []
    for it in items:
        events = events_by_item.get(it.get("name")) or events_by_item.get(it.get("id")) or []
        if not events:
            out.append(it)
            continue
        item_owner = it.get("owner") if it.get("owner") is not None else owner  # nullish, not falsy
        a = attribute_users(events, top_n=top_n, owner=item_owner)
        out.append({
            **it,
            "topUsers": a["topUsers"],
            "userCount": a["userCount"],
            "background": a["background"],
            "owner": a["owner"] if a["owner"] is not None else it.get("owner"),
            "attributionMode": a["mode"],
        })
    return out
